subsys/bluetooth/services/ias/ias_client.c

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2022 Codecoup
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <stdint.h>
+#include <zephyr/sys/check.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/services/ias.h>
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_IAS_CLIENT)
+#define LOG_MODULE_NAME bt_ias_client
+#include "../../common/log.h"
+
+enum {
+	IAS_DISCOVER_IN_PROGRESS,
+
+	IAS_NUM_FLAGS, /* keep as last */
+};
+
+struct bt_ias_client {
+	/* Handle for alert writes */
+	uint16_t alert_level_handle;
+
+	/** Internal flags **/
+	ATOMIC_DEFINE(flags, IAS_NUM_FLAGS);
+
+	/* Gatt discover procedure parameters */
+	struct bt_gatt_discover_params discover;
+};
+
+static const struct bt_uuid *alert_lvl_uuid = BT_UUID_ALERT_LEVEL;
+static const struct bt_uuid *ias_uuid = BT_UUID_IAS;
+
+static const struct bt_ias_client_cb *ias_client_cb;
+static struct bt_ias_client client_list[CONFIG_BT_MAX_CONN];
+
+static struct bt_ias_client *client_by_conn(struct bt_conn *conn)
+{
+	return &client_list[bt_conn_index(conn)];
+}
+
+static void client_cleanup(struct bt_ias_client *ias_client)
+{
+	(void)memset(ias_client, 0, sizeof(*ias_client));
+}
+
+static void discover_failed(struct bt_conn *conn, int err)
+{
+	BT_DBG("conn %p", (void *)conn);
+
+	client_cleanup(client_by_conn(conn));
+	ias_client_cb->discover(conn, err);
```

Comment:
Should NULL-check the callback before calling it:
```suggestion
	if (ias_client_cb != NULL && ias_client_cb->discover != NULL) {
		ias_client_cb->discover(conn, err);
	}
```

Reply:
Done